bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieStateManager.java

```diff
@@ -0,0 +1,310 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package org.apache.bookkeeper.bookie;
+
+import static org.apache.bookkeeper.bookie.BookKeeperServerStats.SERVER_STATUS;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.io.IOException;
+import java.net.UnknownHostException;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.bookkeeper.conf.ServerConfiguration;
+import org.apache.bookkeeper.discover.RegistrationManager;
+import org.apache.bookkeeper.discover.ZKRegistrationManager;
+import org.apache.bookkeeper.stats.Gauge;
+import org.apache.bookkeeper.stats.NullStatsLogger;
+import org.apache.bookkeeper.stats.StatsLogger;
+import org.apache.bookkeeper.util.DiskChecker;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An implementation of StateManager.
+ */
+public class BookieStateManager implements StateManager {
+    private static final Logger LOG = LoggerFactory.getLogger(BookieStateManager.class);
+    private final ServerConfiguration conf;
+    private final LedgerDirsManager ledgerDirsManager;
+
+
+    // use an executor to execute the state changes task
+    final ExecutorService stateService = Executors.newSingleThreadExecutor(
+            new ThreadFactoryBuilder().setNameFormat("BookieStateManagerService-%d").build());
+
+    // Running flag
+    private volatile boolean running = false;
```

Comment:
After I investigated the original code, I found that the state of a bookie has several dimensions, which means the states has overlap, such as the read/writable and running/shuttingdown, for example, a bookie which is shuttingdown will has the readOnly state. The legacy state expression is confusing, a more clear state is needed. So what potential impact of the new state should be investigated first.

Reply:
I don't like the idea to couple logic refactoring with code refactoring. I would suggest not changing anything when doing an abstraction refactor. It is much clearer to reason about when we move code around. 

If there is a need to refactor the logic, this should be a separate PR in a separate discussion after code refactor. 

@ivankelly

.

Reply:
@ArvinDevel 
>  a bookie which is shuttingdown will has the readOnly state

What is the readOnly state in this case, though? It's a flag to check whether you can write to the bookie, i.e. that it's not in the WRITABLE state, which is what we really should be checking.

@sijie 
> If there is a need to refactor the logic, this should be a separate PR in a separate discussion after code refactor.

Sure, this can go in as is and the states can be cleaned up later.